### PR TITLE
Add support to Import/Export table to/from dataframe

### LINKED ISSUE #97: Export/Import from a dataframe

### PR DESCRIPTION
**Pull request by Agent-Hellboy**
fixes  #97

### CODE REVIEW

**Comment by Agent-Hellboy** [comment]
Sorry, for adding multiple commits, You can squash while merging. I tried but I was not able to squash these. I am not able to figure out why the formatting is failing.

File: beautifultable/beautifultable.py [hunk 1]
```diff
@@ -30,6 +30,9 @@
 import copy
 import csv
 import warnings
+from numpy import r_
```
Comment (pri22296) [inline beautifultable/beautifultable.py:33]: why is this import needed? don't see it getting used.

File: beautifultable/beautifultable.py [hunk 2]
```diff
@@ -30,6 +30,9 @@
...
+
+import pandas as pd
```
Comment (pri22296) [inline beautifultable/beautifultable.py:35]: Can we make `pandas` an optional dependency? so ideally just raise an exception inside the `to_df`, `from_df` methods if pandas is not installed. do make them required dependencies for the tests though. I think setup.py has fields for this scenario.

File: beautifultable/beautifultable.py [hunk 3]
```diff
@@ -1213,3 +1216,47 @@ def from_csv(self, file_name, header=True, **kwargs):
             for row in csv_reader:
                 self.rows.append(row)
             return self
+
+    def to_df(self):
+        """Export table to dataframe.
+
+        Parameters
+        ----------
+        table : Beautifuletable
```
Comment (pri22296) [inline beautifultable/beautifultable.py:1225]: typo

File: beautifultable/beautifultable.py [hunk 4]
```diff
@@ -1213,3 +1216,47 @@ def from_csv(self, file_name, header=True, **kwargs):
...
+        Raises
+        ------
+        AttributeError
+            If header is not there in table
+        """
+
+        headers = self.columns.header
+        headers = list(headers)
+        if not headers:
+            raise AttributeError("Dataframe needs a header")
```
Comment (pri22296) [inline beautifultable/beautifultable.py:1236]: do you think a ValueError might be more appropriate here?

File: beautifultable/beautifultable.py [hunk 5]
```diff
@@ -1213,3 +1216,47 @@ def from_csv(self, file_name, header=True, **kwargs):
...
+
+        # index will act as a row header for beautifultable
+        index = list(self.rows.header)
+        mk_dict = {header: self.columns[header] for header in headers}
```
Comment (pri22296) [inline beautifultable/beautifultable.py:1240]: this will lead to unnecessary instantiations. why not use the `data` attribute of the table to get the list of rows. and then use that to generate a dataframe.

File: test.py [hunk 6]
```diff
@@ -5,12 +5,15 @@
...
 
+import pandas as pd
+
 from beautifultable import BeautifulTable
 
 
 class TableOperationsTestCase(unittest.TestCase):
     def setUp(self):
         self.create_table()
+        self.create_dataframe()
```
Comment (pri22296) [inline test.py:16]: wouldn't this mean that `create_dataframe` would get called before all the unit tests even unrelated ones? maybe generate the dataframes within the specifc tests

File: test.py [hunk 7]
```diff
@@ -802,6 +808,16 @@ def test_csv_import(self):
         # Teardown step.
         os.remove("beautiful_table.csv")
 
+    def test_df__export(self):
```
Comment (pri22296) [inline test.py:811]: double underscores

File: test.py [hunk 8]
```diff
@@ -802,6 +808,16 @@ def test_csv_import(self):
...
+        df = self.table.to_df()
+        self.assertEqual(self.table.rows.header, df.index)
+        self.assertEqual(self.table.columns.header, list(df.columns))
+
+    def test_df_import(self):
+        table = BeautifulTable()
+        table = table.from_df(self.df)
+        self.assertEqual(self.table.rows.header, self.df.index)
```
Comment (pri22296) [inline test.py:819]: we should also compare the cell values, not just the headers. also maybe add other unit tests to check the behavior in case headers are missing.

**Review by pri22296 — CHANGES_REQUESTED** [review]

**Comment by pri22296** [comment]
@Agent-Hellboy added few comments. btw the formatting tests are failing because you haven't run `black` on the source code. It's not documented properly that it should be done. apologies for that.

https://github.com/psf/black

File: beautifultable/beautifultable.py (hunk 1 above)
Comment (Agent-Hellboy) [reply beautifultable/beautifultable.py]: Removed.

File: beautifultable/beautifultable.py (hunk 2 above)
Comment (Agent-Hellboy) [reply beautifultable/beautifultable.py]: Sure, will look into this.

File: beautifultable/beautifultable.py (hunk 3 above)
Comment (Agent-Hellboy) [reply beautifultable/beautifultable.py]: Resolved

File: beautifultable/beautifultable.py (hunk 4 above)
Comment (Agent-Hellboy) [reply beautifultable/beautifultable.py]: I realized just now that pandas pick default values for columns and indexes.

File: beautifultable/beautifultable.py (hunk 5 above)
Comment (Agent-Hellboy) [reply beautifultable/beautifultable.py]: Done

File: test.py (hunk 6 above)
Comment (Agent-Hellboy) [reply test.py]: Done

File: test.py (hunk 7 above)
Comment (Agent-Hellboy) [reply test.py]: Done

File: test.py (hunk 8 above)
Comment (Agent-Hellboy) [reply test.py]: Done

File: beautifultable/beautifultable.py (hunk 3 above)
Comment (pri22296) [reply beautifultable/beautifultable.py]: It seems you changed it to **BeautifullTable**. it should be **BeautifulTable**

File: beautifultable/beautifultable.py (hunk 4 above)
Comment (pri22296) [reply beautifultable/beautifultable.py]: I see that you had to check if all members are null to identify the case of no headers. there should be a simpler way for that. I'll keep this in mind.

File: test.py (hunk 6 above)
Comment (pri22296) [reply test.py]: Let's not use the `_data` attribute in the unit tests. private variable so only meant to be used from internal modules.
`[list(row) for row in list(self.table._data)]` => `[list(row) for row in table.rows]`

File: test.py (hunk 8 above)
Comment (pri22296) [reply test.py]: Let's not use the _data attribute in the unit tests. private variable so only meant to be used from internal modules.
[list(row) for row in list(self.table._data)] => [list(row) for row in table.rows]

Comment (Agent-Hellboy) [reply test.py]: Yes, I have changed this in beautifultable.py also. do you know I used `[list(row) for row in table.rows]` while adding asdict API, don't know why I used  _data there.
No probs, I am thinking of picking the row-span, column-span, and transpose features there I will get some good exposure to the design.

**Comment by Agent-Hellboy** [comment]
Thanks for the quick review Priyam 
I don't know why formatting is failing for test.py. I ran black and the failure message is suggesting that everything is okay.
```
 proshan@proshan-ThinkPad-L480  ~/ADD_FEATURE_BEAUTIFULTABLE/beautifultable   master  black test.py 
All done! ✨ 🍰 ✨
1 file left unchanged.
```
Am I missing some flag while running `black test.py`?
one more question When are you planning to add the type annotation?

**Comment by pri22296** [comment]
@Agent-Hellboy Looks like formatting on master is also failing. I did test locally with the latest version of `black` and I do see the errors. Make sure you're using python3 to install and run `black`.

so somethink like

```
python3 -m pip install black
python3 -m black test.py
```

About annotations, can you open an issue for that. Better to track that in a separate thread.

File: test.py (hunk 8 above)
Comment (pri22296) [reply test.py]: spans might be tricky. needs some design before any implementations. I did start with something but found some gaps in the approach hence had to abandon the idea. but yeah, would love to see if you come up with an approach on how to implement these on their respective threads.

Comment (Agent-Hellboy) [reply test.py]: Sure thing

**Comment by Agent-Hellboy** [comment]
@pri22296  I got why formatting was failing. str literals are Unicode instead of bytes in 3.9, so I think we don't need to use u"" anymore. I was using black with 3.8. when I used the latest python nightly 3.11dev, it formatted test.py correctly.
Now CI is failing with PyPy.

**Comment by Agent-Hellboy** [comment]
still failing for PyPy, Here people are saying in comments that `we can't build pandas in pypy without a hack`
https://stackoverflow.com/questions/39500612/any-way-to-install-pandas-with-pypy

File: beautifultable/beautifultable.py [hunk 9]
```diff
@@ -1213,3 +1213,53 @@ def from_csv(self, file_name, header=True, **kwargs):
...
+
+        # If there are column headers then it will act as a column of datafarme
+        headers = list(self.columns.header)
+        if headers.count(None) == len(headers):
+            headers = None
+
+        # If there are row headers then it will act as an Index
+        index = list(self.rows.header)
+        if index.count(None) == len(index):
+            index = None
```
Comment (pri22296) [inline beautifultable/beautifultable.py:1238]: Have we considered the case where some values in the headers are `None` and some aren't? what happens then? pandas wouldn't allow duplicate indexes so would it raise an exception?

File: beautifultable/beautifultable.py [hunk 10]
```diff
@@ -1213,3 +1213,53 @@ def from_csv(self, file_name, header=True, **kwargs):
...
+
+        # Index of dataframe will act as a row headers
+        row_header = list(df.index)
+
+        for header in headers:
+            self.columns.append(
+                [data[header][indx] for indx in row_header], header=str(header)
+            )
+        if list(range(len(row_header))) != row_header:
+            self.rows.header = row_header
```
Comment (pri22296) [inline beautifultable/beautifultable.py:1264]: in case of reading from a dataframe, let's just use the indexes as row headers as is. so, by default, let that be [0,1,2,...]. If the user wants to hide them, they'll add an explicit call `table.rows.headers = None` after importing the dataframe.

**Review by pri22296 — CHANGES_REQUESTED** [review]

**Comment by pri22296** [comment]
> @pri22296 are you waiting for me to increase the code coverage, or you are busy in your work?

Hi, apologies for the late reply. I can see that there are still few unresolved comments. could you please address those so we can go ahead with this change.

**Comment by Agent-Hellboy** [comment]
> > @pri22296 are you waiting for me to increase the code coverage, or you are busy in your work?
> 
> @Agent-Hellboy apologies for the late reply. I can see that there are still few unresolved comments. could you please address those so we can go ahead with this change.

are you talking about these 2
https://github.com/pri22296/beautifultable/pull/126#discussion_r842051779
https://github.com/pri22296/beautifultable/pull/126#discussion_r842054477

**Comment by pri22296** [comment]
@Agent-Hellboy That's so weird. I don't see any comment from you in that thread. Can you maybe paste your comments here? must be some issue on github's end.

**Comment by Agent-Hellboy** [comment]
>Have we considered the case where some values in the headers are None and some aren't? what happens then? pandas wouldn't allow duplicate indexes so would it raise an exception? I'm not sure if it even allows None as an index value.

Actually, they are allowing these things but later giving some intelligent response by taking some decision
example 
it is allowing the duplicate index
```
>>> df.rank()
   rank  rank  rank
0   2.0   1.0   1.0
1   1.0   2.0   2.0
>>> df.to_dict()
<stdin>:1: UserWarning: DataFrame columns are not unique, some columns will be omitted.
{'rank': {0: 'boy', 1: 'girl'}}
```
It is allowing None as the header also
```
>>> df.columns
Index([None, 'p', 'rank'], dtype='object')
```

>in case of reading from a dataframe, let's just use the indexes as row headers as is. so, by default, let that be [0,1,2,...]. If the user wants to hide them, they'll add an explicit call table.rows.headers = None after importing the dataframe.
Done
